Algorithm.Framework/Portfolio/MeanVarianceOptimizationPortfolioConstructionModel.py

```diff
@@ -0,0 +1,165 @@
+            minimum_weight(float): The lower bounds on portfolio weights
+            maximum_weight(float): The upper bounds on portfolio weights
+            target_return(float): The target portfolio return
+            optimization_method(method): Method used to compute the portfolio weights
+        """
+        self.resolution = kwargs['resolution'] if 'resolution' in kwargs else Resolution.Daily
+        self.period = kwargs['period'] if 'period' in kwargs else 63
+        self.minimum_weight = kwargs['minimum_weight'] if 'minimum_weight' in kwargs else -1
+        self.maximum_weight = kwargs['maximum_weight'] if 'maximum_weight' in kwargs else 1
+        self.target_return = kwargs['target_return'] if 'target_return' is kwargs else 2
+        self.optimization_method = kwargs['optimization_method'] if 'optimization_method' in kwargs else self.minimum_variance
+        self.symbolDataBySymbol = {}
+
+
+    def CreateTargets(self, algorithm, insights):
+        """ 
+        Create portfolio targets from the specified insights
+        Args:
+            algorithm: The algorithm instance
+            insights: The insights to create portoflio targets from
+        Returns: 
+            An enumerable of portoflio targets to be sent to the execution model
+        """
+        symbols = [insight.Symbol for insight in insights]
+        if len(symbols) == 0:
+            return []
+
+        for insight in insights:
+            symbolData = self.symbolDataBySymbol.get(insight.Symbol)
+            symbolData.Add(algorithm.Time, insight.Magnitude)
+
+        # Create a dictionary keyed by the symbols in the insights with an pandas.Series as value to create a data frame
+        returns = { str(symbol) : data.Return for symbol, data in self.symbolDataBySymbol.items() if symbol in symbols }
+        returns = pd.DataFrame(returns)
+
+        # The optimization method processes the data frame 
+        opt, weights = self.optimization_method(returns)
+
+        # Create portfolio targets from the specified insights
+        for insight in insights:
+            weight = weights[str(insight.Symbol)]
+            yield PortfolioTarget.Percent(algorithm, insight.Symbol, weight)
+
+
+    def OnSecuritiesChanged(self, algorithm, changes):
+        '''Event fired each time the we add/remove securities from the data feed
+        Args:
+            algorithm: The algorithm instance that experienced the change in securities
+            changes: The security additions and removals from the algorithm'''
+
+        # clean up data for removed securities
+        removed = [ x.Symbol for x in changes.RemovedSecurities ]
+        if len(removed) > 0:
+            for subscription in algorithm.SubscriptionManager.Subscriptions:
+                symbol = subscription.Symbol
+                if symbol in removed and symbol in self.symbolDataBySymbol:
+                    self.symbolDataBySymbol.pop(symbol)
```

Comment:
When removing this symbol we should also remove the consolidator so that the memory can be cleared. To do this, we'll need to keep track of the consolidator instance, can use `algorithm.ResolveConsolidator` to get one that matches the configured resolution/primary tick type, and then can call `algorithm.SubscriptionManager.RemoveConsolidator(symbol, consolidator)` to clear out the events allowing the garbage collector to get rid of the attached indicators and consolidator.